File: algorithms/leetcode/maximum_average_subarray_1.go
package main
import "fmt"
func sum(nums []int) int {
  s := 0
  for _, num := range nums {
    s +=num
  }
  return s
}
func max(a, b int) int {
  if a > b {
    return a
  }
  return b
}
func findMaxAverage(nums []int, k int) float64 {
  max_sum := sum(nums[0:k])
  current_sum := max_sum
  left := 1
  right := k 
  for right < len(nums) {
    current_sum = current_sum - nums[left-1] + nums[right]
    max_sum = max(current_sum, max_sum)
    left +=1
    right +=1
  }
  return float64(max_sum) / float64(k)
}
func main() {

  fmt.Println(findMaxAverage([]int{1,12,-5,-6,50,3}, 4))
}

File: algorithms/leetcode/maximum_average_subarray_1.rs
fn sum(nums: Vec<i32>) -> i32 {
    let mut s = 0;
    for num in nums.iter() {
        s += num;
    }
    return s;
}
fn max(a: i32, b: i32) -> i32 {
    if a > b {
        return a;
    }
    return b;
}

fn find_max_average(nums: Vec<i32>, k: i32) -> f64 {
    let mut max_sum = sum(nums[0..k as usize].to_vec());
    let mut current_sum = max_sum;
    let mut left: usize = 1;
    let mut right: usize = k as usize;

    while right < nums.len() {
        current_sum = current_sum - nums[left - 1] + nums[right];
        max_sum = max(current_sum, max_sum);
        left += 1;
        right += 1;
    }
    return max_sum as f64 / k as f64;
}
fn main() {
    let res = find_max_average(vec![1, 12, -5, -6, 50, 3], 4);
    println!("{}", res);
}
